### PR TITLE
Level0 and Container Documentation

### PR DESCRIPTION
**Pull request by jleidel**
Markdown files for Level0 (basic introduction) and Container documentation.  The former contains very basic info on DES/PDES.  The latter contains information regarding how to build/install/run Singularity, Docker and Spack containers.  All files are in markdown syntax compatible with Github.  Please feel free to make comments/suggest updates regarding content, verbiage and other forms.  Level1-3 documentation is forthcoming.  Docker material made generic from @researcherben original content.

### CODE REVIEW

File: Tutorials/Containers.md [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+The above command will modify user mappings set in `/etc/subuid` and group mappings
+in `/etc/subgid`.
+
+Once built, the container functions as normal. The `.def` files in this repo work with
+or without the `fakeroot` switch and can be built using a command similar to the example below:
+```
+singularity build --fakeroot sstcore-11.0.0-centos7.sif sstcore-11.0.0-centos7.def
+```
+
+Verifying that the container was build successfully can be performed using the following:
```
Comment (gvoskuilen) [inline Tutorials/Containers.md:126]: build -> built

File: Tutorials/Containers.md [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+```
+$> singularity exec sstcore-11.0.0-ubuntu-18.04.sif sst --help
+```
+
+#### Advanced: Container Construction <a name="basiccontainerconstruction"></a>
+
+Users also have the ability build and/or modify custom SST container definitions.  
+We *HIGHLY* suggest that you utilize the existing container definitions mentioned 
+above as the basis for modified containers.  The container definitions Apache
```
Comment (gvoskuilen) [inline Tutorials/Containers.md:135]: definitions *are* Apache

File: Tutorials/Containers.md [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+
+```
+%files
+    sstcore-11.0.0.tar.gz /opt/SST/11.0.0/src/
+```
+
+The `%environment` section ensures that the correct environment variables 
+are initialized in the user's environment when the container is executed.  
+Any potential environment variables (shell variables) can be initialized in 
+this section.  AN example of an `%environment` section is as follows:
```
Comment (gvoskuilen) [inline Tutorials/Containers.md:180]: "AN" -> "An"

File: Tutorials/Containers.md [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+CMD ["/sbin/my_init"]
+
+# Update and install packages
+RUN apt update && apt -y upgrade && apt -y install \
+    build-essential \
+    doxygen \
+    libtool-bin \
+    graphviz \
+    time \
+    mpich \
```
Comment (gvoskuilen) [inline Tutorials/Containers.md:519]: There a reason to use mpich vs openmpi?

File: Tutorials/Containers.md [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+
+WORKDIR $dir
+
+# Utilize the package downloaded from the SST Release page
+COPY sstcore-X.Y.Z.tar.gz .
+RUN tar zxvf sstcore-X.Y.Z.tar.gz
+RUN mv sstcore-X.Y.Z sst-core
+
+# Build SST Core
+RUN cd $dir/sst-core && ./autogen.sh && \
```
Comment (gvoskuilen) [inline Tutorials/Containers.md:553]: autogen isn't needed if you build from the release, only if you get SST from the repo.

File: Tutorials/Containers.md [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+* centos8
+
+Prior to building the provided SST containers, you must first bootstrap the 
+environment.  The repository contains a `bootstrap.sh` script that can be 
+utilized to initialize the environment, find available packages and download 
+the appropriate SST source packages.  Bootstrapping the environment can be performed 
+as follows:
+
+```
+$> git clone https://github.com/tactcomplabs/sst-containers.git
```
Comment (bhpayne) [inline Tutorials/Containers.md:88]: if it's possible to not depend on external repos, that would be preferable.

File: Tutorials/Level0Tutorial.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+exchanging messages.  
+
+How is this different from traditional "simulators?"  Traditional 
+simulators often operate with sequential execution models.  For 
+every new iteration of simulated wall clock time, the simulator 
+must consider updates to each and every simulated component.  This 
+often implies very large runtime overheads for even simplistic 
+simulation models.  In contrast, PDES models only update the portions 
+of the simulation model (or models) that register changes (events) 
+within the given simulated window of time. In this manner, A discrete
```
Comment (bhpayne) [inline Tutorials/Level0Tutorial.md]: lower-case: "a discrete"

File: Tutorials/Containers.md [hunk 8]
```diff
@@ -0,0 +1,598 @@
...
+### Building Basic Singularity Containers <a name="buildingbasicsingularitycontainers"></a>
+
+#### Pre-Built Formulas <a name="basicprebuildformulas"></a>
+
+We *HIGHLY* suggest that users utilize the pre-built singularity container 
+formulas.  These formulas contain basic installations of SST-Core and/or SST-Elements
+that can be utilized and/or modified to serve as the basis for a functional 
+SST environment.  The pre-built SST container repository is found below.  
+
+[SST Container Repository](https://github.com/tactcomplabs/sst-containers)
```
Comment (bhpayne) [inline Tutorials/Containers.md:62]: migrating this content into github.com/sstsimulator/ would be desirable

File: Tutorials/Containers.md [hunk 9]
```diff
@@ -0,0 +1,598 @@
...
+You can submit the aforementioned Slurm script to the batch system using the following command:
+```
+sbatch -N8 slurm.sh
+```
+
+### Singularity References <a name="singularityreferences"></a>
+- [Singularity Installation](https://sylabs.io/guides/3.8/user-guide/quick_start.html#quick-installation-steps)
+- [Singularity Networking Options](https://sylabs.io/guides/3.0/user-guide/networking.html)
+- [Singularity Container Definition Docs](https://sylabs.io/guides/3.0/user-guide/definition_files.html)
+- [SST Container Repository](https://github.com/tactcomplabs/sst-containers)
```
Comment (bhpayne) [inline Tutorials/Containers.md:470]: migrating content to github.com/sstsimulator would be desirable

**Review by bhpayne — COMMENTED** [review]
minor observations

File: Tutorials/Containers.md (hunk 6 above)
Comment (jleidel) [reply Tutorials/Containers.md:88]: We can do one of several things:

- merge the sst-containers repo over to the official SST organization
- setup a public container registry for which public users can pull pre-built container images

There are likely other options, so I'm not opposed to exploring further

File: Tutorials/Level0Tutorial.md (hunk 7 above)
Comment (jleidel) [reply Tutorials/Level0Tutorial.md]: fixed in 8461a20

File: Tutorials/Containers.md (hunk 6 above)
Comment (researcherben) [reply Tutorials/Containers.md:88]: @gvoskuilen -- what do you think of having "sst-containers" as part of github.com/sstsimulator ?

Comment (jleidel) [reply Tutorials/Containers.md:88]: We're definitely good with that.  I would prefer to delay that until we have the CentOS{7,8} containers functional post EOL before we migrate things over.

**Pull request merged**